daft/expressions/expressions.py

```diff
@@ -777,6 +777,34 @@ def concat(self, other: str) -> Expression:
         # Delegate to + operator implementation.
         return Expression._from_pyexpr(self._expr) + other
 
+    def extract(self, pattern: str | Expression, index: int = 0) -> Expression:
+        """Extracts the regex match group from each string in a string column. If index is 0, the entire match is returned, otherwise the specified group is returned.
```

Comment:
Also specify the behavior when the pattern does not match or the index is out of bounds